controllers/machine_api_controllers_controller.go

```diff
@@ -0,0 +1,125 @@
+package controllers
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+
+	"github.com/google/go-jsonnet"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	imagesConfigMapName            = "machine-api-operator-images"
+	originalUpstreamDeploymentName = "machine-api-controllers"
+	imageKey                       = "images.json"
+
+	caBundleConfigMapName = "appuio-machine-api-ca-bundle"
+)
+
+//go:embed machine_api_controllers_deployment.jsonnet
+var deploymentTemplate string
+
+// MachineAPIControllersReconciler creates a appuio-machine-api-controllers deployment based on the images.json ConfigMap
+// if the upstream machine-api-controllers does not exist.
+type MachineAPIControllersReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	Namespace string
+}
+
+func (r *MachineAPIControllersReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	if req.Name != imagesConfigMapName {
+		return ctrl.Result{}, nil
+	}
+
+	l := log.FromContext(ctx).WithName("UpstreamDeploymentReconciler.Reconcile")
+	l.Info("Reconciling")
+
+	var imageCM corev1.ConfigMap
+	if err := r.Get(ctx, req.NamespacedName, &imageCM); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	ij, ok := imageCM.Data[imageKey]
+	if !ok {
+		return ctrl.Result{}, fmt.Errorf("images.json key not found in ConfigMap %q", imagesConfigMapName)
```

Comment:
Maybe use `imageKey` in the message for consistency?

```suggestion
		return ctrl.Result{}, fmt.Errorf("%q key not found in ConfigMap %q", imageKey, imagesConfigMapName)
```